### PR TITLE
Add block examples for all core blocks

### LINKED ISSUE #17488: Inserter Panel: Need block previews for each block
With the introduction of the Inserter Help Panel, each time the user hovers over a block, they should see an example of what that block might look like in the panel. Right now we don't include this, but it needs to be implemented for inclusion into WP 5.3.

Let's get this rolling. 
**I'll begin creating sample blocks and exporting the HTML to include here in the comments.**

Screen Shot 2019-09-19 at 10 48 22 AM

### PR DESCRIPTION
**Pull request by youknowriad**
closes #17488

This PR adds block examples for the core blocks in order to show a preview of the blocks in the inserter help panel. It uses the examples in the #17488 issue.

Screenshot 2019-09-23 at 13 07 58

image

### CODE REVIEW

File: packages/block-library/src/rss/index.js [hunk 1]
```diff
@@ -20,5 +20,10 @@ export const settings = {
 		align: true,
 		html: false,
 	},
+	example: {
+		attributes: {
+			feedURL: 'https://wordpress.org',
```
Comment (Soean) [inline packages/block-library/src/rss/index.js]:
```suggestion
			feedURL: __( 'https://wordpress.org' ),
```

**Review by Soean — COMMENTED** [review]
I added some suggestions.

File: packages/block-library/src/rss/index.js (hunk 1 above)
Comment (richtabor) [reply packages/block-library/src/rss/index.js]: > Suggested change 
> 			feedURL: 'https://wordpress.org',
> 			feedURL: __( 'https://wordpress.org' ),

URL's don't need to be translated. 👌

Comment (HardeepAsrani) [reply packages/block-library/src/rss/index.js]: It's usually helpful to translate users to redirect users to the homepage of their locale website like https://hi.wordpress.org/ if they're using WP in that locale.

File: packages/block-library/src/image/index.js [hunk 2]
```diff
@@ -25,6 +25,13 @@ export const settings = {
 		'img', // "img" is not translated as it is intended to reflect the HTML <img> tag.
 		__( 'photo' ),
 	],
+	example: {
+		attributes: {
+			sizeSlug: 'large',
+			url: 'https://images.unsplash.com/photo-1549880339-d93e3072aef4',
+			caption: __( 'Snow covered mountain' ),
```
Comment (gziolo) [inline packages/block-library/src/image/index.js]: Image(s) from Unsplash probably need a copyright note.

File: packages/blocks/src/api/factory.js [hunk 3]
```diff
@@ -461,3 +462,17 @@ export function switchToBlockType( blocks, name ) {
...
+
+/**
+ * Create a block object from the example API.
+ *
+ * @param {string} name
+ * @param {Object} example
+ *
+ * @return {Object} block.
+ */
+export const getBlockFromExample = ( name, example ) => {
```
Comment (gziolo) [inline packages/blocks/src/api/factory.js]: Should it be `createBlockFromExample` for conistency with the default name and the description?
```suggestion
export const createBlockFromExample = ( name, example ) => {
```

File: packages/block-library/src/image/index.js (hunk 2 above)
Comment (mtias) [reply packages/block-library/src/image/index.js]: We should bundle any image we use as well.

**Comment by mtias** [comment]
This is a great place to add some personality.

File: packages/block-library/src/image/index.js (hunk 2 above)
Comment (youknowriad) [reply packages/block-library/src/image/index.js]: Can't we use images hosted on wp.org instead? Bundling could be challenging as the public URL is not easily accessible here.

File: packages/block-library/src/video/index.js [hunk 4]
```diff
@@ -21,6 +21,12 @@ export const settings = {
 	description: __( 'Embed a video from your media library or upload a new one.' ),
 	icon,
 	keywords: [ __( 'movie' ) ],
+	example: {
+		attributes: {
+			src: 'https://make.wordpress.org/design/files/2019/03/block-manager.mp4',
```
Comment (mcsf) [inline packages/block-library/src/video/index.js]: This doesn't play in Safari.

Comment (mtias) [reply packages/block-library/src/video/index.js]: Is there any way to have a very small auto-playing video instead? It doesn't make sense to load a video that you cannot play.

File: packages/block-library/src/columns/index.js [hunk 5]
```diff
@@ -24,8 +24,51 @@ export const settings = {
...
+					{
+						name: 'core/paragraph',
+						attributes: {
+							content: __( 'Nam risus massa, ullamcorper consectetur eros fermentum, porta aliquet ligula. Sed vel mauris nec enim ultricies commodo.' ),
+						},
+					},
+				],
+			},
+		],
+	},
```
Comment (mcsf) [inline packages/block-library/src/columns/index.js]: Needs overflow taken care of.

Captura de ecrã 2019-09-23, às 14 19 05

**Comment by jasmussen** [comment]
This is really nice. Thank you for doing this. 

This marquee: 
Screenshot 2019-09-23 at 16 12 58

:slow clap:

Not sure if it's something for this PR, probably not, but there's a horizontal scrollbar here:

Screenshot 2019-09-23 at 16 12 38

And this is possibly a theme issue (CC: @kjellr for insights):

Screenshot 2019-09-23 at 16 12 24

it's a little low.

Many of those seem to be more specific to the preview component, or the theme styling them. So not a blocker for this PR probably. And this PR is 👌

**Comment by mapk** [comment]
Can the font-sizes get closer to match the Paragraph block preview? Several of the block previews are very small.

* List block
* Preformatted block
* Verse block
* HTML block
* Code block

panel-sizes

**Comment by youknowriad** [comment]
@mapk Preview works by adapting the size to the container. So if the block is wider, it can be shown smaller than the others. It's going to be hard to align the size of all the blocks. We can probably try to tweak a bit.

Are we satisfied with the examples so far? I'd like us to have at least something in, even if not perfect for WordPress 5.3.

Also, should we replace Unsplash photos with other photos without licensing issues and upload them to wp.org?

**Comment by mcsf** [comment]
In reply to https://github.com/WordPress/gutenberg/pull/17493#issuecomment-534119100

> it's a little low.

I think this is due to the loading of the images and not knowing their size beforehand. @jasmussen, if you hover once, move the cursor away, and hover again, does the positioning change?

**Comment by jasmussen** [comment]
@mcsf 

> I think this is due to the loading of the images and not knowing their size beforehand. @jasmussen, if you hover once, move the cursor away, and hover again, does the positioning change?

Short answer, yes. Now I think it's a race condition. I did this:

- Check out the branch
- Open the library
- Mouse over the block
- The content was vertically centered, but as the image "trickled in" it was vertically offset
- Mousing over a different block then back reset it because now the dimensions can be red and accurately positioned.

Perhaps just adding a "height" attribute shoudl fix this.

Because it was a thing about loading, though I wanted to record a GIF of it, so I cleared my cache and started recording a GIF, and now the image won't load at all. Not sure what I did wrong:

preview

**Comment by mapk** [comment]
As for the examples being used...

### Group block 
The preview doesn't make much sense to me.

Screen Shot 2019-09-24 at 10 49 11 AM

### Columns block
Everything else is a bit more creative than simple Lorem Ipsum. Let's replace it with this Robert Frost poem instead.

```
<!-- wp:columns -->
<div class="wp-block-columns"><!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p>Two roads diverged in a yellow wood, And sorry I could not travel both, And be one traveler, long I stood, And looked down one as far as I could, To where it bent in the undergrowth;</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>Then took the other, as just as fair, And having perhaps the better claim, Because it was grassy and wanted wear; Though as for that the passing there, Had worn them really about the same,</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>And both that morning equally lay. In leaves no step had trodden black. Oh, I kept the first for another day! </p>
<!-- /wp:paragraph --></div>
<!-- /wp:column -->

<!-- wp:column -->
<div class="wp-block-column"><!-- wp:paragraph -->
<p>Yet knowing how way leads on to way, I doubted if I should ever come back.</p>
<!-- /wp:paragraph -->

<!-- wp:image {"sizeSlug":"large"} -->
<figure class="wp-block-image size-large"><img src="https://images.unsplash.com/photo-1549880339-d93e3072aef4" alt=""/></figure>
<!-- /wp:image -->

<!-- wp:paragraph -->
<p>I shall be telling this with a sigh, Somewhere ages and ages hence: Two roads diverged in a wood, and I––I took the one less traveled by,<br>And that has made all the difference.</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>– Robert Frost</p>
<!-- /wp:paragraph --></div>
<!-- /wp:column --></div>
<!-- /wp:columns -->
```

**Comment by mtias** [comment]
@mapk it seems you are missing colors:

image

**Comment by mapk** [comment]
I wonder if we should instead do something like this for the Group block?

Screen Shot 2019-09-24 at 12 51 40 PM


```
<!-- wp:group {"backgroundColor":"very-dark-gray"} -->
<div class="wp-block-group has-very-dark-gray-background-color has-background"><div class="wp-block-group__inner-container"><!-- wp:paragraph {"customTextColor":"#ffffff","customBackgroundColor":"#9258d8"} -->
<p style="background-color:#9258d8;color:#ffffff" class="has-text-color has-background">Block one.</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph {"customTextColor":"#ffffff","customBackgroundColor":"#3e8fdc"} -->
<p style="background-color:#3e8fdc;color:#ffffff" class="has-text-color has-background">Block two.</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph {"customTextColor":"#ffffff","customBackgroundColor":"#5bcc88"} -->
<p style="background-color:#5bcc88;color:#ffffff" class="has-text-color has-background">Block three.</p>
<!-- /wp:paragraph --></div></div>
<!-- /wp:group -->
```

**Comment by mtias** [comment]
I tried something very close to that but looked too overwhelming / forced in comparison, so settled with the colored text for now.

**Comment by mapk** [comment]
> I tried something very close to that but looked too overwhelming / forced in comparison, so settled with the colored text for now.

I think you're right with a dark background. Maybe with a lighter background?

Screen Shot 2019-09-26 at 2 53 18 PM

Also, I'm not sure why, but I don't get the colored text in my preview. And it's not registering with me visually as a "group".

**Comment by mapk** [comment]
Or maybe using the text approach, but with a colored background on the Group block...

Screen Shot 2019-09-26 at 2 58 28 PM

**Comment by jasmussen** [comment]
I did not see colored text in twenty nineteen. Perhaps something with them registering custom color palettes.

**Comment by mapk** [comment]
I'm using the Gutenberg Starter Theme, so not sure if that effects it either.

**Comment by youknowriad** [comment]
I think we should not rely on color palettes for the example and only use custom colors because these work across themes.

**Comment by youknowriad** [comment]
I reverted the base64 commit because in my testing it increases the bundle size of the block-library script by 1Mb. Let's use URLs for now until we figure out a better approach to bundle these images in Core instead.

**Comment by youknowriad** [comment]
updated the group example to avoid the use of the color palette. It should work well across themes now.

**Review by jasmussen — COMMENTED** [review]
I really like this. I think we should scramble to get it in. 

On very slow connections it does surface a few improvements we can make to the preview component., but since that's only the first time you hover a block and then never after, I would not call that a blocker. Furthermore, it drastically improves the usability for sighted users of the block library, and it helps encourage 3rd party block developers to include a preview, that it's an important feature to ship.

However I think we should make a quick change. We are hotlinking the images, and we are using Unsplash. The latters license while very liberal, the suggestion to credit photographers makes it perhaps 0.05% less liberal than a strict interpretation of the GPL, which suggests complete and total freedom.

So just in order to be totally kosher, I would recommend replacing all images with certified public domain images from Wikimedia. Using this list https://commons.wikimedia.org/wiki/Featured_Pictures_in_the_Public_Domain I would suggest we could use the following:

- https://commons.wikimedia.org/wiki/File:Windbuchencom.jpg
- https://commons.wikimedia.org/wiki/File:MarsSunset.jpg
- https://commons.wikimedia.org/wiki/File:MarsSunset.jpg
- https://commons.wikimedia.org/wiki/File:M104_ngc4594_sombrero_galaxy_hi-res.jpg
- https://commons.wikimedia.org/wiki/File:Glacial_lakes,_Bhutan.jpg
- https://commons.wikimedia.org/wiki/File:Mount_Hood_reflected_in_Mirror_Lake,_Oregon.jpg
- https://commons.wikimedia.org/wiki/File:Kuznetsk_Alatau_1.jpg
- https://commons.wikimedia.org/wiki/File:Sediment_off_the_Yucatan_Peninsula.jpg
- https://commons.wikimedia.org/wiki/File:Rocks_El_Torcal_de_Antequera_karst_Andalusia_Spain.jpg
- https://commons.wikimedia.org/wiki/File:Icy_shores_of_Lake_Michigan.jpg
- https://commons.wikimedia.org/wiki/File:Drymona_water_fall_and_pool_north_Euboea_Greece.jpg
- https://commons.wikimedia.org/wiki/File:Beech_and_ferns_in_Gullmarsskogen.jpg

An additional benefit is that Wikemedia explicitly permits hotlinking: https://commons.wikimedia.org/wiki/Commons:Reusing_content_outside_Wikimedia/technical

I hope to take a stab at replacing the URLs and returning. 

The overall content, other than the images, are great, though, works well:

examples

**Comment by jasmussen** [comment]
Okay, pushed good fastloading guaranteed publicdomain images that look good. These five, for reference:

- https://commons.wikimedia.org/wiki/File:Biologia_Centrali-Americana_-_Cantorchilus_semibadius_1902.jpg
- https://commons.wikimedia.org/wiki/File:Windbuchencom.jpg
- https://commons.wikimedia.org/wiki/File:Glacial_lakes,_Bhutan.jpg
- https://commons.wikimedia.org/wiki/File:Sediment_off_the_Yucatan_Peninsula.jpg
- https://commons.wikimedia.org/wiki/File:MtBlanc1.JPG

Had to look for Mont Blanc and a specific bird due to the prose around them.

A couple more things before we can merge:

- https://github.com/WordPress/gutenberg/pull/17493/files — commit suggestion?
- https://github.com/WordPress/gutenberg/pull/17493/files#r327064974 — I'm looking for a public domain video that plays in Safari, if you have one let me know

I will hopefully be back.

File: packages/block-library/src/video/index.js (hunk 4 above)
Comment (jasmussen) [reply packages/block-library/src/video/index.js]: So, I found this place for public domain video: https://www.pond5.com/stock-video-footage/1/.html#1/18447/durationlt:10,pricelt:0

This one is cute, albeit short: https://www.pond5.com/stock-footage/44576423/little-smokey-walking-cage.html

But the problem then becomes, can we hotlink it? Where should these things be hosted? Is make.wordpress.org okay?

**Comment by jasmussen** [comment]
I just updated the video to be a very short and generic WordPress logo file. This is more appropriate for the preview where you can't even play it back. Plus the preview looks good:

Screenshot 2019-09-30 at 15 55 37

With the help of @kjellr I'm investigating why it doesn't play back in Safari, as noted in the reviews. It may be the Make server.

**Review by jasmussen — APPROVED** [review]
As noted previously, I really like this. I think we should get it in.

I replaced all the images with Wikimedia images that are public domain. They also allow us to hotlink, so that helps. In addition, the files are small, so they load really fast compared the the previous images, and given they will be shown in scaled down previews the fidelity is fine. 

I also removed the video, because the one we had did not play in Safari, and it should be hosted on a CDN. I created #17655 to track this. 

What remains here is lovely, and I think it's good to go.

However I also did a bit of work on this, so would appreciate a sanity check, for example from @kjellr.

**Comment by mapk** [comment]
Just created a Trac ticket to host images on w.org's CDN: https://core.trac.wordpress.org/ticket/48292

**Comment by hacknug** [comment]
I know it's a minor issue but, could we address the sizing of `.block-editor-block-card` items changing from block to block? It's somewhat annoying trying to preview blocks and read their names and descriptions without having those in a fixed position.

Guessing it's due to flex-grow/shrink since the explicit sizing seems to be in place already.

**Pull request merged**